app/src/main/java/com/dreamisi/moviebase/data/Repository.kt

```diff
@@ -0,0 +1,117 @@
+package com.dreamisi.moviebase.data
+
+import com.dreamisi.moviebase.data.JsonModels.JsonGenreList
+import com.dreamisi.moviebase.data.JsonModels.JsonMovieDetails
+import com.dreamisi.moviebase.data.models.Movie
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import com.dreamisi.moviebase.data.JsonModels.JsonPopularMoviesResponse
+import com.dreamisi.moviebase.data.models.Genre
+import com.dreamisi.moviebase.data.models.MovieDetails
+
+class Repository {
+
+    private var moviesList: List<Movie>? = null
+    private var genreList: List<Genre>? = null
+    private val networkModule = NetworkModule
+    private var movieDetails: MovieDetails? = null
+
+
+    suspend fun getMoviesList(): List<Movie> = withContext(Dispatchers.IO) {
+        val cachedGenre = genreList
+        if (cachedGenre == null) {
+            val genreFromDB = loadGenresFromDB()
+            genreList = genreFromDB
+        }
+        val cachedMovies = moviesList
+        if (cachedMovies != null) {
+            cachedMovies
+        } else {
+            val moviesFromDB = loadMoviesFromDB(genreList ?: emptyList())
+            moviesList = moviesFromDB
+            moviesFromDB
+        }
+    }
+
+    private suspend fun loadGenresFromDB(): List<Genre> {
+        val data = networkModule.theMovieDataBaseAPI.getGenreList()
+        return parseGenre(data)
```

Comment:
Опять же, в kotlin можно попроще делать такие приобразования без дополнительных переменных:
```suggestion
        return networkModule.theMovieDataBaseAPI
            .getGenreList()
            .let { parseGenre(it) }
```

Reply:
Тут конечно мб вкусовщина, но я бы поправил по такому принципу и остальные функции в классе